### PR TITLE
[Applications] Optimize token sampling

### PR DESCRIPTION
**Pull request by haehun**
## Dependency of the PR
None

## Commits to be reviewed in this PR

<details><summary>Faster sampling</summary><br />

Let `n` be number of tokens and `k` be k in top-k sampling.
In current sampling code, the time complexity is O(nlogn) from `std::sort`.
This PR improves time complexity to O(nlogk) with `std::partial_sort`.
In a typical model, where `n ~ 10 ^ 5` and `k ~ 40`, this PR makes significant difference in token generation.

With Qwen2-0.5b model, the improvement is from...

```
=================[ LLM with NNTrainer ]===================
prefill: 8 tokens, 161 ms, 49.6894 TPS
generation: 512 tokens, 50984 ms, 10.0424 TPS
==========================================================
```

to
```
=================[ LLM with NNTrainer ]===================
prefill: 8 tokens, 155 ms, 51.6129 TPS
generation: 512 tokens, 46860 ms, 10.9262 TPS
==========================================================
```

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Haehun Yang <haehun.yang@samsung.com>

</details>

### Summary

- Faster sampling

Signed-off-by: Haehun Yang <haehun.yang@samsung.com>

### CODE REVIEW

File: Applications/LLaMA/jni/main.cpp [hunk 1]
```diff
@@ -94,7 +96,7 @@ float applyTKP(float *logits, int len, float temperature, unsigned int top_k,
   }
 
   // Apply Top-K and Top-P
-  std::fill_n(logits, sizeof(len), -INFINITY);
+  std::fill_n(logits, len, -INFINITY);
```
Comment (songgot) [inline Applications/LLaMA/jni/main.cpp:99]: The bug fix is ​​great. :+1:

**Review by songgot — COMMENTED** [review]
Great work on optimizing the token sampling performance with std::partial_sort and fixing the critical initialization bug in std::fill_n! :+1:

File: Applications/CausalLM/llm_util.cpp [hunk 2]
```diff
@@ -79,8 +81,10 @@ float applyTKP(float *logits, int len, float temperature, unsigned int top_k,
       logits[i] = logits[i] / temperature;
     top_indices_and_logits.push_back({i, logits[i]});
   }
-  sort(top_indices_and_logits.begin(), top_indices_and_logits.end(),
-       [](auto &a, auto &b) { return a.second > b.second; });
+  std::partial_sort(top_indices_and_logits.begin(),
+                    top_indices_and_logits.begin() + 1,
+                    top_indices_and_logits.end(),
+                    [](auto &a, auto &b) { return a.second > b.second; });
```
Comment (jaemini-shin) [inline Applications/CausalLM/llm_util.cpp:88]: Not this PR, it seems like that this function doesn't apply top-K, P on logits but returns the max logit only.
I think it should be the same logic in the `LLaMA/jni/main.cpp`.

Comment (haehun) [reply Applications/CausalLM/llm_util.cpp:88]: I agree that top-K and P in this file should be fixed, but the fix can be in separate PR.
